File: src/core/ClusterModel.cpp
#include "ClusterModel.hpp"
#include "Simulation.hpp"

void ClusterModel::combine(const ::std::vector<UserCell*>& cells, const ::std::vector<UserMoleculeSpace*>&)
{
    size_t length = cells.size();
    //細胞同士が十分に近ければくっついたと判定する
    for (size_t i = 0; i < length; ++i) {
        UserCell& cell = *cells[i];
        switch (cell.getCellType()) {
            case CellType::WORKER:
            {
                for (size_t j = i + 1; j < length; j++) {

                    UserCell& cellr = *cells[j];

                    if (cellr.getCellType() != CellType::WORKER) continue;
        
                    //
                    double radius = cell.getRadius() + cellr.getRadius();
        
                    //距離 < 細胞半径 + 細胞半径 でくっつく
                    //高速化のため、２乗で計算
                    if ((cell.getPosition() - cellr.getPosition()).squareLength() < radius * radius) {
                        cell.combine(cellr);
                        continue;
                    }
                }
                break;
            }

            default: break;
        }
        
    }
}

void ClusterModel::combine(const ::std::vector<UserCell*>& cells, const ::std::vector<UserMoleculeSpace*>& moleculeSpaces, CellList& cellList)
{
    size_t length = cells.size();
    //細胞同士が十分に近ければくっついたと判定する
    for (size_t i = 0; i < length; ++i) {
        UserCell& cell = *cells[i];
        CellInfo info = CellInfo(cell);

        switch (info.cellType) {
            case CellType::WORKER:
            {
                for (CellInfo cellInfo : cellList.iterateAffectableCellInfos(cell, cells, moleculeSpaces)) {

                    if (cellInfo.cellType != CellType::WORKER) continue;
        
                    //
                    double radius = info.radius + cellInfo.radius;
        
                    //距離 < 細胞半径 + 細胞半径 でくっつく
                    //高速化のため、２乗で計算
                    if ((info.position - cellInfo.position).squareLength() < radius * radius) {
                        cell.combine(*info.pCell);
                        continue;
                    }
                }
                break;
            }

            default: break;
        }
        
    }
}

void ClusterModel::combine(const ::std::vector<UserCell*>& cells, const ::std::vector<UserMoleculeSpace*>& moleculeSpaces, CellList* pCellList)
{
    if (pCellList == nullptr) combine(cells, moleculeSpaces);
    else combine(cells, moleculeSpaces, *pCellList);
}

File: src/cs/CellSim/CellAlgorithms/ClusterModel.cs
using CellSim.Cells;
using CellSim.Molecular;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CellSim.CellAlgorithms
{
    public static class ClusterModel
    {
        public static void Combine(List<Cell> cells, IReadOnlyList<ReadOnlyMoleculeField> fields)
        {
            Combine(cells, fields, null);
        }

        public static void Combine(List<Cell> cells, IReadOnlyList<ReadOnlyMoleculeField> fields, CellList? cellList)
        {
            static void f(Cell cell, CellInfo info, CellInfo cellInfo)
            {
                if (cellInfo.Type == CellType.Invalid) return;
                if (cellInfo.Type != info.Type) return;

                double radius = info.Radius + cellInfo.Radius;

                //距離 < 細胞半径 + 細胞半径 でくっつく
                //高速化のため、２乗で計算
                if ((info.Position - cellInfo.Position).SquareLength < radius * radius)
                {
#pragma warning disable CS8604 // Null 参照引数の可能性があります。
                    cell.Combine(cellInfo.Source as Cell);
#pragma warning restore CS8604 // Null 参照引数の可能性があります。
                }
            }

            if (cellList is null)
            {
                for (int i = 0; i < cells.Count; i++)
                {
                    Cell cell = cells[i];
                    CellInfo info = new CellInfo(cell);
                    for (int j = i + 1; j < cells.Count; j++)
                    {
                        f(cell, info, new CellInfo(cells[j]));
                    }
                }
            }
            else
            {
                foreach (Cell cell in cells)
                {
#pragma warning disable CS8625 // null リテラルを null 非許容参照型に変換できません。
                    CellAlgorithm.Enumerate
                    (
                        null,
                        cellList,
                        cell,
                        cells,
                        fields,
                        cellInfo => f(cell, new CellInfo(cell), cellInfo)
                    );
#pragma warning restore CS8625 // null リテラルを null 非許容参照型に変換できません。
                }
                
            }
        }
    }
}
